### PR TITLE
Thread Infrastructure Code Review

### PR DESCRIPTION
**Pull request by MalinkyZubr**
Removed pycache and added to gitignore. 

Dynamic net score calculation
Multiprocessing staging + running infrastructure
class structure
92% test coverage

### CODE REVIEW

File: src/metric.py [hunk 1]
```diff
@@ -1,74 +1,282 @@
...
-    priority: int
-
 
     def __init__(self):
+        """
+        Initializes the BaseMetric with default values.
+        """
         self.score: float = 0.0
+        self.url: str = ""
+        self.priority = 1
```
Comment (silica-dev) [inline src/metric.py:29]: needs type annotation

File: src/metric.py [hunk 2]
```diff
@@ -1,74 +1,282 @@
...
 
-class ModelMetadata(BaseModel):
-    model_url: typing.Optional[str]
-    code_url: typing.Optional[str]
-    dataset_url: typing.Optional[str]
+class ModelURLs(BaseModel):
+    """
+    Stores URLs related to a model, including model, codebase, and dataset URLs.
+    """
+    model: typing.Optional[str]
```
Comment (silica-dev) [inline src/metric.py:12]: my IDE is saying this type was deprecated in python 3.10 and to use `str | None` instead. idk how much this matters though.

File: src/metric.py [hunk 3]
```diff
@@ -1,74 +1,282 @@
...
+    def __init__(self, base_coefficient: int):
+        """
+        Initializes the exponential decay function.
+        Args:
+            base_coefficient (int): The base coefficient (> 1).
+        """
+        assert(base_coefficient > 1)
+        self.base_coefficient: int = base_coefficient
+
+    def calculate_priority_weight(self, priority: int) -> float:
```
Comment (silica-dev) [inline src/metric.py:119]: should be marked @typing.override

File: src/metric.py [hunk 4]
```diff
@@ -1,74 +1,282 @@
...
+            float: The calculated net score.
+        """
+        num_metrics: int = len(metrics)
+
+        priority_organized_scores: SortedDict[int, list[float]] = self.generate_scores_priority_dict(metrics)
+        compressed_scores: list[list[float]] = self.compress_priorities(priority_organized_scores)
+        priority_weights: list[float] = self.get_priority_weights(compressed_scores, num_metrics)
+        aggregated_scores: list[float] = [self.sum_scores(scores) for scores in compressed_scores]
+
+        net_score: float = sum(
```
Comment (silica-dev) [inline src/metric.py:178]: im seeing warnings about partially unknown types here, but I don't know the best way to sort them out

File: src/workflow.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
+from multiprocessing import Pool
+from typing import Literal, Optional
+from pydantic import BaseModel
+import typing
+from metric import BaseMetric, ModelURLs, AnalyzerOutput, PRIORITY_FUNCTIONS, PriorityFunction
```
Comment (silica-dev) [inline src/workflow.py:5]: unused import of PriorityFunction

File: tests/infrastructure.py
Comment (silica-dev) [inline tests/infrastructure.py:1]: tests seem good to me so far, though I do agree that more general ones could be good. one thing not tested is the edge case of someone setting a metric to priority zero, which might be important since this is meant to be modular

File: dependencies.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+pydantic==1.8.2
+sortedcontainers==2.4.0
+poetry==1.1.11
```
Comment (silica-dev) [inline dependencies.txt:3]: is there a reason we're using such an old version of poetry? it caused a few issues on my system.

**Review by silica-dev — CHANGES_REQUESTED** [review]
Overall I think this is structurally correct, there's just a few tweaks I have regarding code style, tests, and dependencies.

File: src/metric.py (hunk 3 above)
Comment (MalinkyZubr) [reply src/metric.py:119]: added

File: src/metric.py (hunk 1 above)
Comment (MalinkyZubr) [reply src/metric.py:29]: addes

**Comment by MalinkyZubr** [comment]
2 additional tests added for net score calculation

**Review by silica-dev — APPROVED** [review]
